substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/graal/code/SubstrateBackend.java

```diff
@@ -152,14 +166,220 @@ public CompilationResult newCompilationResult(CompilationIdentifier compilationI
         return new CompilationResult(compilationIdentifier, name) {
             @Override
             public void close(OptionValues options) {
-                /*
-                 * Do nothing, we do not want our CompilationResult to be closed because we
-                 * aggregate all data items and machine code in the native image heap.
-                 */
+                FrameOptimizer.compressInfoPoints(this);
             }
         };
     }
 
+    private static class FrameOptimizer {
+        /**
+         * Wrapper class that makes up for deficiencies in the implementation of
+         * BytecodeFrame.equals.
+         */
+        private static class BytecodeFrameComparator {
+            private BytecodeFrame frame;
+
+            BytecodeFrameComparator(BytecodeFrame frame) {
+                assert frame != null;
+                this.frame = frame;
+            }
+
+            @Override
+            public int hashCode() {
+                return frame.hashCode();
+            }
+
+            @Override
+            public boolean equals(Object obj) {
+                if (this == obj) {
+                    return true;
+                }
+                if (obj == null || !(obj instanceof BytecodeFrameComparator)) {
+                    return false;
+                }
+                BytecodeFrameComparator that = (BytecodeFrameComparator) obj;
+                BytecodeFrame thisFrame = frame;
+                BytecodeFrame thatFrame = that.frame;
+                while (thisFrame != thatFrame && thisFrame != null && thatFrame != null) {
+                    if (thisFrame.getBCI() == thatFrame.getBCI() &&
+                                    Objects.equals(thisFrame.getMethod(), thatFrame.getMethod()) &&
+                                    Objects.equals(thisFrame.caller(), thatFrame.caller()) &&
+                                    // BytecodeFrame.equals returns true even when these tests fail!
```

Comment:
@dougxc is fixing this in https://bugs.openjdk.org/browse/JDK-8300590

If you find JVMCI bugs like this, please let us know instead of working around the problem. We could have had the proper fix in all the just released updates of the JDK already.

Reply:
Hi @christianwimmer. Apologies for not raising this as a bug. I had assumed that there was a good reason for disabling this check in the subtype implementation. In future I'll attempt to curb my natural assumption that the GraalVM team must know what they are doing! ;-)